### PR TITLE
xgboost: 0.90 -> 1.4.1

### LINKED ISSUE #114426: xgboost is out of date
###### Checklist

- [X] Checked the [nixpkgs master branch](https://github.com/NixOS/nixpkgs)

- [X] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

###### Project name
`nix search` name: xgboost

current version: 0.90
desired version: 1.3.3

###### Notify maintainers

maintainers: @abbradar

###### Note for maintainers

Please tag this issue in your PR.

### PR DESCRIPTION
**Pull request by rmcgibbo**
###### Motivation for this change
https://nix-cache.s3.amazonaws.com/log/yq7j62wggafqf7zfb9a2rydaknx463qx-python3.8-xgboost-0.90.drv
ZHF: #122042
Closes: #114426, #63094

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/xgboost/default.nix [hunk 1]
```diff
@@ -1,44 +1,58 @@
...
                 ++ lib.optional ncclSupport nccl;
 
-  cmakeFlags = lib.optionals cudaSupport [ "-DUSE_CUDA=ON" "-DCUDA_HOST_COMPILER=${cudatoolkit.cc}/bin/cc" ]
-               ++ lib.optional ncclSupport "-DUSE_NCCL=ON";
+  cmakeFlags = [
+    "-DGOOGLE_TEST=ON"
+  ] ++ lib.optionals cudaSupport [ "-DUSE_CUDA=ON" "-DCUDA_HOST_COMPILER=${cudatoolkit.cc}/bin/cc" ]
+    ++ lib.optionals ncclSupport [ "-DUSE_NCCL=ON" ];
+
+  doCheck = true;
```
Comment (jonringer) [inline pkgs/development/libraries/xgboost/default.nix:38]: may want to make this into an input, (e.g. `, doCheck ? true`) then optionally bring testing dependencies.

Awkward to force tests to be built even if `doCheck = false;`

File: pkgs/development/python-modules/xgboost/default.nix [hunk 2]
```diff
@@ -10,28 +10,52 @@
...
+    hypothesis
   ];
 
-  postPatch = "cd python-package";
+  # Override existing logic for locating libxgboost.so which is not appropriate for Nix
+  prePatch = let
+    libPath = "${xgboost}/lib/libxgboost${stdenv.hostPlatform.extensions.sharedLibrary}";
+  in ''
+    echo 'find_lib_path = lambda: ["${libPath}"]' > python-package/xgboost/libpath.py
+  '';
```
Comment (jonringer) [inline pkgs/development/python-modules/xgboost/default.nix:38]: hmmm cc @FRidh 

How do you feel about this type of patching

I could see issues in the future if the filename changes... but those would also be issues with the current patching paradigm as well

Comment (rmcgibbo) [reply pkgs/development/python-modules/xgboost/default.nix:38]: The prior patch file stopped working when I updated to 1.4.1, and it basically just removed this entire file anyways: https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/libpath.py#L14-L65. I figured this was easier for future maintainers than creating a new patch file.

File: pkgs/development/libraries/xgboost/default.nix (hunk 1 above)
Comment (rmcgibbo) [reply pkgs/development/libraries/xgboost/default.nix:38]: Is that the standard for C++-y packages? I wasn't sure. I don't think these tests are _really_ necessary, but in "python land" I usually try to turn on tests when cleaning up a package, if they're currently off. I was trying to do the same here.

Comment (jonringer) [reply pkgs/development/libraries/xgboost/default.nix:38]: I agree completely :)

But I could see a scenario in the future in which the test suite is having issues, and being able to disable them would be nice; especially for overriding.

File: pkgs/development/python-modules/xgboost/default.nix (hunk 2 above)
Comment (jonringer) [reply pkgs/development/python-modules/xgboost/default.nix:38]: Yea, I can see the merits and appeal of doing it this way. But fridh has much more historical context, and would like his opinion; if he's okay with it, I think we can move forward.

**Comment by rmcgibbo** [comment]
Result of `nixpkgs-review pr 122234` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>python38Packages.dask-xgboost</li>
    <li>python38Packages.optuna</li>
    <li>python39Packages.dask-xgboost</li>
    <li>python39Packages.optuna</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>haskellPackages.xgboost-haskell</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>python38Packages.xgboost</li>
    <li>python39Packages.xgboost</li>
    <li>xgboost</li>
  </ul>
</details>

------------------

```
error: builder for '/nix/store/z957s8dfkxp2pk7d9jcgls4wppqifdsf-xgboost-haskell-0.1.0.0.drv' failed with exit code 1;
       last 10 log lines:
       > impure path `/private/tmp/nix-build-xgboost-haskell-0.1.0.0.drv-0/ghc21628_0/ghc_26.filelist' used in link
       >
       > <no location info>: error:
       >     `ld' failed in phase `Merge objects'. (Exit code: 1)
       > [4 of 5] Compiling ML.DMLC.XGBoost.Rabit.FFI ( src/ML/DMLC/XGBoost/Rabit/FFI.hs, dist/build/ML/DMLC/XGBoost/Rabit/FFI.o, dist/build/ML/DMLC/XGBoost/Rabit/FFI.dyn_o )
       > impure path `/private/tmp/nix-build-xgboost-haskell-0.1.0.0.drv-0/ghc21628_0/ghc_18.filelist' used in link
       >
       > <no location info>: error:
       >     `ld' failed in phase `Merge objects'. (Exit code: 1)
       > 
       For full logs, run 'nix log /nix/store/z957s8dfkxp2pk7d9jcgls4wppqifdsf-xgboost-haskell-0.1.0.0.drv'.
```

**Review by jonringer — APPROVED** [review]
LGTM

Result of nixpkgs-review pr 122234 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)<details>  <summary>4 packages marked as broken and skipped:</summary>  <ul>    <li>python38Packages.dask-xgboost</li>    <li>python38Packages.optuna</li>    <li>python39Packages.dask-xgboost</li>    <li>python39Packages.optuna</li>  </ul></details><details>  <summary>4 packages built:</summary>  <ul>    <li>haskellPackages.xgboost-haskell</li>    <li>python38Packages.xgboost</li>    <li>python39Packages.xgboost</li>    <li>xgboost</li>  </ul></details>

**Pull request merged**